File: frontend/src/types/match.ts
import { formatKickoff } from '@/utils/datetime'

export type MatchStatus = 'scheduled' | 'upcoming' | 'live' | 'finished' | 'postponed' | 'cancelled'

export type Stage =
  | 'group'
  | 'group_stage'
  | 'round_of_32'
  | 'round_of_16'
  | 'quarter_final'
  | 'semi_final'
  | 'third_place'
  | 'final'
  | 'regular_season'

/** Backend raw match object */
export interface ApiMatch {
  id: number
  match_no: number
  stage: Stage
  group_id: number | null
  group: { id: number; name: string } | null
  competition_id?: number | null
  season?: number | null
  matchday?: number | null
  home_team_id: number
  away_team_id: number
  home_team: { id: number; name: string; name_en: string; fifa_code: string | null; external_code?: string | null; flag_url: string }
  away_team: { id: number; name: string; name_en: string; fifa_code: string | null; external_code?: string | null; flag_url: string }
  home_score: number | null
  away_score: number | null
  status: MatchStatus
  live_minute?: number | null
  kickoff_time_utc: string
  stadium: { id: number; name: string; city_id: number } | null
  city: { id: number; name: string } | null
  importance_level: number
  recommend_tag: string
  recommend_reason: string
  winner_team_id: number | null
  created_at: string
  updated_at: string
}

/** Frontend normalized match */
export interface Match {
  id: number
  match_number: number
  stage: Stage
  group_id?: number | null
  group_name: string | null
  home_team_id: number
  away_team_id: number
  home_team_name: string
  away_team_name: string
  home_team_code: string
  away_team_code: string
  home_flag: string
  away_flag: string
  home_score: number | null
  away_score: number | null
  status: MatchStatus
  kickoff_time_utc: string
  local_kickoff_time: string
  city: string
  stadium: string
  importance_level: number
  is_featured: boolean
  minute: number | null
  matchday: number | null
  competition_id: number | null
}

/** Transform backend match to frontend match */
export function normalizeMatch(m: ApiMatch): Match {
  return {
    id: m.id,
    match_number: m.match_no,
    stage: m.stage,
    group_id: m.group_id ?? null,
    group_name: m.group?.name || null,
    home_team_id: m.home_team_id,
    away_team_id: m.away_team_id,
    home_team_name: m.home_team?.name || '',
    away_team_name: m.away_team?.name || '',
    home_team_code: m.home_team?.fifa_code || m.home_team?.external_code || '',
    away_team_code: m.away_team?.fifa_code || m.away_team?.external_code || '',
    home_flag: m.home_team?.flag_url || '',
    away_flag: m.away_team?.flag_url || '',
    home_score: m.home_score,
    away_score: m.away_score,
    status: m.status,
    kickoff_time_utc: m.kickoff_time_utc,
    local_kickoff_time: formatKickoff(m.kickoff_time_utc),
    city: m.city?.name || '',
    stadium: m.stadium?.name || '',
    importance_level: m.importance_level,
    is_featured: m.importance_level >= 2,
    minute: m.live_minute ?? null,
    matchday: m.matchday ?? null,
    competition_id: m.competition_id ?? null,
  }
}

// Unified match status mapping (DATA-04): every card/detail page must use
// these helpers so postponed/cancelled/live-minute semantics stay consistent.
export function matchStatusLabel(status: string, minute: number | null): string {
  if (status === 'live') {
    // minute may be null when the provider has no live clock data:
    // never render a fake "0'".
    return minute != null ? `${minute}'` : '直播中'
  }
  if (status === 'finished') return '已结束'
  if (status === 'postponed') return '已延期'
  if (status === 'cancelled') return '已取消'
  return '未开始'
}

export function matchStatusClass(status: string): string {
  if (status === 'live') return 'live'
  if (status === 'finished') return 'green'
  if (status === 'postponed' || status === 'cancelled') return 'warn'
  return ''
}

File: backend/internal/models/match.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Match struct {
	ID               uint           `gorm:"primaryKey" json:"id"`
	MatchNo          int            `gorm:"not null" json:"match_no"`
	HomeTeamID       uint           `gorm:"index" json:"home_team_id"`
	AwayTeamID       uint           `gorm:"index" json:"away_team_id"`
	HomeTeam         Team           `gorm:"foreignKey:HomeTeamID" json:"home_team,omitempty"`
	AwayTeam         Team           `gorm:"foreignKey:AwayTeamID" json:"away_team,omitempty"`
	GroupID          *uint          `gorm:"index" json:"group_id"`
	Group            *Group         `gorm:"foreignKey:GroupID" json:"group,omitempty"`
	Stage            string         `gorm:"size:30;default:group;index" json:"stage"`
	CompetitionID    *uint          `gorm:"index" json:"competition_id"` // nullable: NULL = World Cup / legacy data
	Season           *int           `gorm:"index" json:"season"`         // starting year, e.g. 2025 for 2025-26 season
	Matchday         *int           `gorm:"index" json:"matchday"`       // league round, league competitions only
	StadiumID        uint           `gorm:"index" json:"stadium_id"`
	Stadium          Stadium        `gorm:"foreignKey:StadiumID" json:"stadium,omitempty"`
	CityID           uint           `gorm:"index" json:"city_id"`
	City             City           `gorm:"foreignKey:CityID" json:"city,omitempty"`
	KickoffTimeUTC   time.Time      `gorm:"not null;index" json:"kickoff_time_utc"`
	HomeScore        *int           `json:"home_score"`
	AwayScore        *int           `json:"away_score"`
	Status           string         `gorm:"size:20;default:scheduled;index" json:"status"`
	StatusDetail     string         `gorm:"size:30" json:"status_detail"`
	LiveMinute       *int           `json:"live_minute"`
	WinnerTeamID     *uint          `json:"winner_team_id"`
	ExternalProvider string         `gorm:"size:30;index" json:"external_provider"`
	ExternalID       string         `gorm:"size:64;index" json:"external_id"`
	LastSyncedAt     *time.Time     `json:"last_synced_at"`
	ImportanceLevel  int            `gorm:"default:0" json:"importance_level"`
	RecommendTag     string         `gorm:"size:50" json:"recommend_tag"`
	RecommendReason  string         `gorm:"size:255" json:"recommend_reason"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"-"`
}
